Background-Substractor-Video-Capturer/main.cpp

```diff
@@ -0,0 +1,328 @@
+#include <atomic>
+#include <cstdint>
+#include <cstdlib>
+#include <iostream>
+#include <vector>
+
+#include <opencv2/opencv.hpp>
+#include <opencv2/imgproc.hpp>
+#include <opencv2/video/background_segm.hpp>
+
+#include <opentok.h>
+
+#include "config.h"
+#include "otk_thread.h"
+#include "renderer.h"
+
+static std::atomic<bool> g_is_connected(false);
+static otc_publisher *g_publisher = nullptr;
+static std::atomic<bool> g_is_publishing(false);
+
+struct custom_video_capturer {
+  const otc_video_capturer *video_capturer;
+  struct otc_video_capturer_callbacks video_capturer_callbacks;
+  int format;
+  int width;
+  int height;
+  int fps;
+  otk_thread_t capturer_thread;
+  std::atomic<bool> capturer_thread_exit;
+};
+
+static otk_thread_func_return_type capturer_thread_start_function(void *arg) {
+  struct custom_video_capturer *video_capturer = static_cast<struct custom_video_capturer *>(arg);
+  if (video_capturer == nullptr) {
+    otk_thread_func_return_value;
+  }
+
+  cv::VideoCapture opencv_video_capture(0);
+  if (!opencv_video_capture.isOpened()) {
+    otk_thread_func_return_value;
+  }
+
+  // The following arguments have the default values according to
+  // https://docs.opencv.org/master/de/de1/group__video__motion.html.
+  int history = 500;
+  double var_threshold = 16;
+  bool detect_shadows = true;
+  cv::Ptr<cv::BackgroundSubtractor> background_subtractor =
+      cv::createBackgroundSubtractorMOG2(history, var_threshold, detect_shadows);
+  cv::Mat opencv_frame, foreground_mask, background_image, foreground_image;
+  while((video_capturer->capturer_thread_exit.load() == false) &&
+        opencv_video_capture.read(opencv_frame)) {
+    if (foreground_mask.empty()) {
+      foreground_mask.create(opencv_frame.size(), opencv_frame.type());
+    }
+    // See https://docs.opencv.org/master/d7/d7b/classcv_1_1BackgroundSubtractorMOG2.html
+    // for more info about the learing rate argument.
+    background_subtractor->apply(opencv_frame, foreground_mask, /*learning_rate*/ -1);
```

Comment:
If you created a const variable for the learning_rate, then code will be more self-documenting, IMO.